Adds Flow / propTypes support

tisho opened the pull request:
**Don't panic, most of the changes are in test files.**

This adds Flow and plain propTypes support to the component importer. It does this by breaking up the operation of the importer into separate modes (set by the `--mode` CLI argument):
- `--mode typescript` uses the same TypeScript analyzer as before. No changes there.
- `--mode flow` uses `react-docgen` to analyze source files
- `--mode plain` uses `react-docgen` to analyze source files

Both `flow` and `plain` mode use the exact same analyzer / propType extractor, so they should work fine even in repos where you have a mix of flow-typed + regular propType components. The main difference between them is how they put together the list of files for analysis. I've intentionally aliased the flow and plain analyzer to the same function, but kept the codepaths different in case we see a need for them to diverge more in the future.

Other notable changes and limitations:
- The `flow` and `plain` mode support `defaultProps`.
- Imported types and spreads are supported (forked `react-docgen` under the company account).
- Some of the more exotic `export` test cases are not handled by the new analyzer. This is because in plain/flow mode we do a much more exhaustive search for files that could potentially contain components. We can probably add support for them down the line if we see the importer fail frequently in those cases.

fhur commented on src/analyze/flow.ts:
```diff
@@ -0,0 +1,125 @@
+    }
+}
+
+function toComponentInfo(info): ComponentInfo {
```
Can you add a type here? If its not available from react-docgen can we approximate the type to the information that we expect from it?

fhur commented on src/analyze/flow.ts:
```diff
@@ -0,0 +1,125 @@
+}
+
+function toComponentInfo(info): ComponentInfo {
+    // console.log(util.inspect(info, false, null, true))
```
drop?

fhur commented on src/analyze/handlers/exportedNameHandler.ts:
```diff
@@ -0,0 +1,33 @@
+import * as types from "ast-types"
+import { getNameOrValue } from "../../utils"
+
+const { namedTypes: t } = types
```
What's this? Id you just want to import the `namedTypes`, consider using the more explicit `import { namedTypes as t } from "types"`

tisho replied:
Will do. I was initially using other stuff from `ast-types` and forgot to refactor the references in the end.

fhur commented on src/analyze/handlers/exportedNameHandler.ts:
```diff
@@ -0,0 +1,33 @@
+ * @param path The AST node that's being processed
+ * @param parser AST parser instance
+ */
+export const exportedNameHandler = (documentation, path, parser) => {
```
types :) if not easily available can we approximate them

fhur commented on src/cli/commandGenerateComponents.ts:
```diff
@@ -8,20 +8,25 @@ type CliGenerateComponentsArguments = {
+    [force]       : (optional) Re-writes component files, even if they already exist.
+                    Default: false.
+    [help]        : (optional) Prints this message.
+    [verbose]     : (optional) Increase verbosity of printed out messages.
```
Default: false?

fhur commented on src/cli/commandGenerateConfiguration.ts:
```diff
@@ -59,12 +51,20 @@ Where:
+    [tsconfig]     : (optional) A path to the "tsconfig.json" when in "typescript" mode.
     [help]         : (optional) Prints this message.
-    [force]        : (optional) Will override the existing components.
+    [force]        : (optional) Will overwrite the existing components and configuration.
```
Default: false?

fhur commented on src/cli/commandGenerateConfiguration.ts:
```diff
@@ -59,12 +51,20 @@ Where:
     [help]         : (optional) Prints this message.
-    [force]        : (optional) Will override the existing components.
+    [force]        : (optional) Will overwrite the existing components and configuration.
+    [verbose]      : (optional) Increase verbosity of printed out messages.
```
Default: false?

fhur commented on src/cli/commandGenerateConfiguration.ts:
```diff
@@ -108,30 +126,79 @@ export async function commandGenerateConfiguration(argv: string[]) {
 
-    if (!index || index.length == 0) {
-        printInvalidIndex()
+    let components
```
Will probably need to import the type as well:

```suggestion
    let components: ComponentInfo[]
```

fhur commented on src/cli/commandGenerateConfiguration.ts:
```diff
@@ -108,30 +126,79 @@ export async function commandGenerateConfiguration(argv: string[]) {
-    if (!index || index.length == 0) {
-        printInvalidIndex()
+    let components
+    let result
```
Will probably need to import the type as well:

```suggestion
    let result: EmitResult[]
```

fhur commented on src/compilerOptions.ts:
```diff
@@ -16,8 +16,13 @@ const keyValidator = (key: string, predicate: (x: any) => boolean, message: stri
-    keyValidator("rootFiles", t => Array.isArray(t) && t.length > 0, "Expected 'rootFiles' to be a non empty array"),
+    keyValidator(
+        "rootFiles",
+        t => t === undefined || (Array.isArray(t) && t.length > 0),
```
So we allow an undefined `rootFiles` now?

tisho replied:
`rootFiles` becomes optional when `mode` isn't `typescript`. What would be the best way to handle this? Pass in all other props to the `keyValidator`?

fhur replied:
I see the issue. Or have a set of validators per `mode`, I'll add something.

fhur replied:
Improved this plus some tests for the compilerOptions. Let me know what you think.

tisho replied:
I like it. Thanks!

fhur commented on src/types.ts:
```diff
@@ -40,6 +40,12 @@ export type ComponentConfiguration = {
 export type CompileOptions = {
+    /**
+     * The mode the importer will operate in.
+     * Can be one of `typescript`, `flow`, or `plain` (`propTypes`). Defaults to `typescript`.
```
If it defaults, then this field can be made optional :)

fhur commented on src/utils.ts:
```diff
@@ -1,9 +1,64 @@
+    semi: false,
+}
+
+const { namedTypes: t } = types
```
Same comment here. The `import` syntax is cleaner in my opinion.

fhur commented on src/utils.ts:
```diff
@@ -46,6 +100,78 @@ export async function createPrettierFormatter(prettierrc?: string): Promise<Form
+ * Otherwise, a relative path from the `outDir` to the original path will be generated.
+ */
+
+export function resolveComponentImportPath(importPath, outDir) {
```
types :)

fhur commented on tsconfig.json:
```diff
@@ -5,7 +5,7 @@
         "rootDir": "src",
         "outDir": "build",
-        "strict": true,
+        "strict": false,
```
I can't say I like this. If too strict for your taste, consider enabling/disabling a subset of the checks instead (https://www.typescriptlang.org/docs/handbook/compiler-options.html)

tisho replied:
You're right to doubt this. I don't like it, either. I'll try to make it work with `strict: true`.

tisho replied:
I restored `strict: true` and added a bunch of types back in, including the `react-docgen` ones. Found a reasonable approximation since `react-docgen` actually has Flow types.

fhur commented:
Another thing I just realized: If you look at the [commit messages](https://github.com/framer/component-importer/commits/master) I use the convention "label: short description". This I then use to generate the changelogs, so it be great if you can stick to the convention.

Examples:

fix: fixed a bug
docs: changed the docs
refactor: didn't do anything useful from the user's perspective
feat: added a feature

fhur commented on src/generate/conversion.ts:
```diff
@@ -54,7 +54,7 @@ function createPropertyControlExpression(prop: PropType): ts.ObjectLiteralExpres
 
     // PropertyControl#defaultValue
-    if (prop.defaultValue) {
+    if ("defaultValue" in prop) {
```
Note that the property might be defined, but the value could be undefined. I'm not sure if this is introducing a bug here.

<img width="274" alt="Screenshot 2019-11-11 at 16 42 45" src="https://user-images.githubusercontent.com/6452323/68600322-963cc700-04a2-11ea-886e-1cd924e0b5ac.png">

tisho replied:
It already was a bug, actually. Since it's checking if `defaultValue` is falsy, it would skip the assignment if `defaultValue` was `0`, or an empty string, or an actual boolean `false`. That said, it's probably fine to filter out `undefined` and `null`, specifically, because right now we get a lot of defaults that are `null`, because of the code above and the corresponding defaults override in `heuristics.ts`. I'll work on improving both.

tisho commented on src/__tests__/integration.salesforce.ts:
```diff
@@ -0,0 +1,244 @@
+
+describe("integration-test", () => {
+    test("salesforce lightning", async () => {
+        const importPath = `integration-test-data/@salesforce/design-system-react/node_modules/@salesforce/design-system-react/components`
```
I'll add in some tests against the generated propTypes like in the grommet example.

fhur replied:
I'll add them myself 👍

tisho commented:
> Another thing I just realized: If you look at the [commit messages](https://github.com/framer/component-importer/commits/master) I use the convention "label: short description". This I then use to generate the changelogs, so it be great if you can stick to the convention.
> 
> Examples:
> 
> fix: fixed a bug
> docs: changed the docs
> refactor: didn't do anything useful from the user's perspective
> feat: added a feature

I didn't realize this until it was too late. I'm sorry. Maybe I could just squash all commits before we merge this into `master`, and then write the correct message? Would that work?

tisho commented:
Thanks so much for the feedback @fhur! I hit the re-request button a little too early, so maybe give it another day before you review the changes. Here's what's left to do:

- [ ] Test for specific props in integration tests
- [ ] Improve `defaultValue` handling by ignoring `undefined | null` values
- [ ] Handle extracted component naming conflicts (could pull this into a different PR if need be)
- [ ] Update README and other documentation

fhur commented:
> Handle extracted component naming conflicts

That can be handled separately since it's already a bug in the existing impl. 

> Maybe I could just squash all commits before we merge this into master, and then write the correct message? Would that work?

Squashing before committing to master is my preferred approach 👍 

Do you want me to help you finish any of this?

tisho commented:
@fhur I've completed the outstanding todos, so I think we're good to merge this in and then start the name conflict resolution work in a separate branch. Let me know if you're good with that and I'll squash / merge / release a version. Or if you want to go ahead and do that, that's fine by me, too.
